File: backend/app/schemas/procedure.py
"""Procedure request/response schemas."""

from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field, field_validator


class MaterialUsed(BaseModel):
    """A material used during a procedure."""

    name: str = Field(..., min_length=1, max_length=200)
    quantity: int = Field(default=1, ge=1)
    lot_number: str | None = None


class ProcedureCreate(BaseModel):
    """Fields required to record a new procedure."""

    cups_code: str = Field(..., min_length=4, max_length=10)
    cups_description: str = Field(..., min_length=1, max_length=500)
    tooth_number: int | None = None
    zones: dict | None = None
    materials_used: list[MaterialUsed] | None = None
    duration_minutes: int | None = Field(default=None, ge=1, le=480)
    notes: str | None = None
    treatment_plan_item_id: str | None = None
    clinical_record_id: str | None = None

    @field_validator("cups_code")
    @classmethod
    def validate_cups_format(cls, v: str) -> str:
        import re
        if not re.match(r"^[0-9]{6}$", v.strip()):
            raise ValueError("Código CUPS inválido. Formato esperado: 6 dígitos.")
        return v.strip()


class ProcedureResponse(BaseModel):
    """Full procedure detail."""

    model_config = ConfigDict(from_attributes=True)

    id: str
    patient_id: str
    doctor_id: str
    cups_code: str
    cups_description: str
    tooth_number: int | None = None
    zones: dict | None = None
    materials_used: dict | None = None
    duration_minutes: int | None = None
    notes: str | None = None
    treatment_plan_item_id: str | None = None
    clinical_record_id: str | None = None
    is_active: bool
    created_at: datetime
    updated_at: datetime


class ProcedureListResponse(BaseModel):
    """Cursor-paginated list of procedures."""

    items: list[ProcedureResponse]
    next_cursor: str | None = None
    has_more: bool

File: frontend/lib/validations/procedure.ts
/**
 * Zod validation schemas for procedure forms.
 *
 * These mirror the backend Pydantic schemas for clinical procedure creation.
 * All field names use snake_case to match the backend API.
 *
 * Key validation rules (from CLAUDE.md security spec):
 * - CUPS code: ^[0-9]{6}$ (6 digits)
 * - FDI tooth number: integer 11–85
 */

import { z } from "zod";

// ─── Material Schema ──────────────────────────────────────────────────────────

/**
 * Schema for a single material used during a procedure.
 */
export const procedureMaterialSchema = z.object({
  name: z
    .string()
    .min(1, "El nombre del material es requerido")
    .max(200, "El nombre del material no puede exceder 200 caracteres")
    .transform((v) => v.trim()),
  quantity: z
    .number({ invalid_type_error: "La cantidad debe ser un número" })
    .positive("La cantidad debe ser mayor a 0")
    .optional(),
});

export type ProcedureMaterial = z.infer<typeof procedureMaterialSchema>;

// ─── Procedure Create Schema ──────────────────────────────────────────────────

/**
 * Schema for creating a new clinical procedure record.
 * Mirrors the backend ProcedureCreate Pydantic schema.
 */
export const procedureCreateSchema = z.object({
  cups_code: z
    .string()
    .min(1, "El código CUPS es requerido")
    .regex(/^[0-9]{6}$/, "Código CUPS inválido (debe tener exactamente 6 dígitos)")
    .transform((v) => v.trim()),

  cups_description: z
    .string()
    .min(1, "La descripción del procedimiento es requerida")
    .max(500, "La descripción no puede exceder 500 caracteres")
    .transform((v) => v.trim()),

  tooth_number: z
    .number({ invalid_type_error: "El número de diente debe ser un número" })
    .int("El número de diente debe ser un entero")
    .min(11, "Número de diente inválido (mín. 11)")
    .max(85, "Número de diente inválido (máx. 85)")
    .optional()
    .nullable(),

  zones: z
    .array(z.string().min(1).max(50))
    .max(10, "No puedes seleccionar más de 10 zonas")
    .optional()
    .nullable(),

  materials_used: z
    .array(procedureMaterialSchema)
    .max(50, "No puedes registrar más de 50 materiales")
    .optional()
    .nullable(),

  treatment_plan_item_id: z
    .string()
    .uuid("ID de ítem de plan de tratamiento inválido")
    .optional()
    .nullable(),

  clinical_record_id: z
    .string()
    .uuid("ID de registro clínico inválido")
    .optional()
    .nullable(),

  duration_minutes: z
    .number({ invalid_type_error: "La duración debe ser un número" })
    .int("La duración debe ser un número entero")
    .min(1, "La duración mínima es 1 minuto")
    .max(480, "La duración máxima es 480 minutos (8 horas)")
    .optional()
    .nullable(),
});

export type ProcedureCreate = z.infer<typeof procedureCreateSchema>;
